src/parquet/column/statistics.h

```diff
@@ -0,0 +1,180 @@
+    null_count = value;
+    has_null_count = true;
+    return *this;
+  }
+
+  inline EncodedStatistics& set_distinct_count(int64_t value) {
+    distinct_count = value;
+    has_distinct_count = true;
+    return *this;
+  }
+};
+
+template <typename DType>
+class PARQUET_EXPORT TypedRowGroupStatistics;
+
+class PARQUET_EXPORT RowGroupStatistics
+    : public std::enable_shared_from_this<RowGroupStatistics> {
+ public:
+  int64_t null_count() const { return statistics_.null_count; }
+  int64_t distinct_count() const { return statistics_.distinct_count; }
+  int64_t num_values() const { return num_values_; }
+
+  virtual bool HasMinMax() const = 0;
+  virtual void Reset() = 0;
+  virtual void Merge(const RowGroupStatistics& other) = 0;
+
+  template <typename TypedStats>
+  std::shared_ptr<TypedStats> as();
+
+  // Plain-encoded minimum value
+  virtual std::string EncodedMin() = 0;
+
+  // Plain-encoded maximum value
+  virtual std::string EncodedMax() = 0;
+
+  virtual EncodedStatistics Encode() = 0;
+
+  virtual ~RowGroupStatistics() {}
+
+  Type::type physical_type() const { return descr_->physical_type(); }
+
+ protected:
+  const ColumnDescriptor* descr() const { return descr_; }
+  void SetDescr(const ColumnDescriptor* schema) { descr_ = schema; }
+
+  void IncrementNullCount(int64_t n) { statistics_.null_count += n; }
+
+  void IncrementNumValues(int64_t n) { num_values_ += n; }
+
+  void IncrementDistinctCount(int64_t n) { statistics_.distinct_count += n; }
+
+  void MergeCounts(const RowGroupStatistics& other) {
+    this->statistics_.null_count += other.statistics_.null_count;
+    this->statistics_.distinct_count += other.statistics_.distinct_count;
+    this->num_values_ += other.num_values_;
+  }
+
+  void ResetCounts() {
+    this->statistics_.null_count = 0;
+    this->statistics_.distinct_count = 0;
+    this->num_values_ = 0;
+  }
+
+  const ColumnDescriptor* descr_ = nullptr;
+  int64_t num_values_ = 0;
+  EncodedStatistics statistics_;
+};
+
+template <typename DType>
+class PARQUET_EXPORT TypedRowGroupStatistics : public RowGroupStatistics {
+ public:
+  using T = typename DType::c_type;
+
+  TypedRowGroupStatistics(
+      const ColumnDescriptor* schema, MemoryAllocator* allocator = default_allocator());
+
+  TypedRowGroupStatistics(const T& min, const T& max, int64_t num_values,
+      int64_t null_count, int64_t distinct_count);
+
+  TypedRowGroupStatistics(const ColumnDescriptor* schema, const std::string& encoded_min,
+      const std::string& encoded_max, int64_t num_values, int64_t null_count,
+      int64_t distinct_count, MemoryAllocator* allocator = default_allocator());
+
+  bool HasMinMax() const override;
+  void Reset() override;
+  void Merge(const RowGroupStatistics& other) override;
+
+  void Update(const T* values, int64_t num_not_null, int64_t num_null);
+
+  const T& min() const { return min_; }
+  const T& max() const { return max_; }
+
+  std::string EncodedMin() override;
```

Comment:
Renamed to `EncodeMin`/`EncodeMax` to reflect better that these values are not stored anywhere.